src/rust/pipeline-ingress/src/main.rs

```diff
@@ -0,0 +1,175 @@
+    },
+    pipeline::{
+        v1beta1::{
+            Metadata,
+            RawLog,
+        },
+        v1beta2::Envelope,
+    },
+};
+use thiserror::Error;
+use tokio::net::TcpListener;
+use tracing_subscriber::{
+    prelude::*,
+    EnvFilter,
+};
+use uuid::Uuid;
+
+#[non_exhaustive]
+#[derive(Debug, Error)]
+enum IngressApiError {
+    #[error("failed to send message to kafka {0}")]
+    ProducerError(#[from] ProducerError),
+}
+
+struct IngressApi {
+    producer: Producer<Envelope<RawLog>>,
+}
+
+impl IngressApi {
+    fn new(producer: Producer<Envelope<RawLog>>) -> Self {
+        IngressApi { producer }
+    }
+}
+
+#[async_trait::async_trait]
+impl PipelineIngressApi<IngressApiError> for IngressApi {
+    async fn publish_raw_log(
+        &self,
+        request: PublishRawLogRequest,
+    ) -> Result<PublishRawLogResponse, IngressApiError> {
+        let created_time = SystemTime::now();
+        let last_updated_time = created_time;
+        let tenant_id = request.tenant_id;
+        let event_source_id = request.event_source_id;
+        // TODO: trace_id should be generated at the edge. This service is
+        // currently "the edge" but that won't be true forever. When there is an
+        // actual edge service, that service should be responsible for
+        // generating the trace_id.
+        let trace_id = Uuid::new_v4();
+        self.producer
+            .send(Envelope::new(
+                Metadata::new(
+                    tenant_id,
+                    trace_id,
+                    0,
+                    created_time,
+                    last_updated_time,
+                    event_source_id,
+                ),
+                RawLog::new(request.log_event),
+            ))
+            .await?;
+
+        Ok(PublishRawLogResponse::ok())
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, Error)]
+enum ConfigurationError {
+    #[error("failed to configure kafka client {0}")]
+    Kafka(#[from] KafkaConfigurationError),
+
+    #[error("failed to configure gRPC server {0}")]
+    Server(#[from] ServerConfigurationError),
+
+    #[error("missing environment variable {0}")]
+    EnvironmentVariable(#[from] VarError),
+
+    #[error("failed to bind socket address {0}")]
+    SocketError(#[from] std::io::Error),
+
+    #[error("failed to parse integer value {0}")]
+    ParseInt(#[from] ParseIntError),
+}
+
+#[tracing::instrument(err)]
+async fn handler() -> Result<(), ConfigurationError> {
+    let socket_address = std::env::var("PIPELINE_INGRESS_BIND_ADDRESS")?;
+    let healthcheck_polling_interval_ms =
+        std::env::var("PIPELINE_INGRESS_HEALTHCHECK_POLLING_INTERVAL_MS")?.parse()?;
+
+    tracing::info!("configuring kafka producer");
+    let producer: Producer<Envelope<RawLog>> = Producer::new("raw-logs")?;
+    tracing::info!("kafka producer configured successfully");
+
+    tracing::info!("configuring gRPC server");
+    let (server, _shutdown_tx) = PipelineIngressServer::new(
```

Comment:
`shutdown_tx` would be used once we implement a signal handler to ensure graceful shutdown. It is present but ignored here because when I wrote it `let (server, _) = ...` immediately dropping the `shutdown_tx` caused the server to shutdown immediately. So it must be present, but for now we don't use it.